src/main/java/org/truffleruby/core/cast/DurationToMillisecondsNode.java

```diff
@@ -65,12 +67,25 @@ protected long durationNil(Nil duration) {
     }
 
     @Specialization
-    protected long duration(RubyDynamicObject duration) {
+    protected long duration(RubyDynamicObject duration,
+            @Cached DispatchNode divmodNode,
+            @Cached DispatchNode sumNode,
+            @Cached ToLongNode toLongNode) {
         if (floatCastNode == null) {
             CompilerDirectives.transferToInterpreterAndInvalidate();
             floatCastNode = insert(NumericToFloatNodeGen.create());
         }
-        return duration(floatCastNode.executeDouble(duration));
+        try {
+            return duration(floatCastNode.executeDouble(duration));
+        } catch (Exception e) {
+            try {
+                Object result = divmodNode.call(duration, "divmod", 1);
+                Object out = sumNode.call(result, "sum");
+                return toLongNode.execute(out);
```

Comment:
This node returns a number of milliseconds, so we need to multiply the result by 1000.